pkg/reconciler/instances/eventing/preaction/removenatsoperatorstep.go

```diff
@@ -0,0 +1,109 @@
+package preaction
+
+import (
+	"github.com/kyma-incubator/reconciler/pkg/reconciler/chart"
+	"github.com/kyma-incubator/reconciler/pkg/reconciler/instances/eventing/log"
+	"github.com/kyma-incubator/reconciler/pkg/reconciler/kubernetes"
+	"github.com/kyma-incubator/reconciler/pkg/reconciler/service"
+	"go.uber.org/zap"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"strings"
+)
+
+const (
+	removeNatsOperatorStepName = "removeNatsOperator"
+	natsOperatorLastVersion    = "1.24.7"
+	natsSubChartPath           = "eventing/charts/nats"
+	eventingNats               = "eventing-nats"
+	oldConfigValue             = "global.image.repository"
+	newConfigValue             = "eu.gcr.io/kyma-project"
+)
+
+var (
+	natsOperatorCRDsToDelete = []string{"natsclusters.nats.io", "natsserviceroles.nats.io"}
+)
+
+type kubeClientProvider func(context *service.ActionContext, logger *zap.SugaredLogger) (kubernetes.Client, error)
+
+type removeNatsOperatorStep struct {
+	kubeClientProvider
+}
+
+func newRemoveNatsOperatorStep() *removeNatsOperatorStep {
+	return &removeNatsOperatorStep{
+		kubeClientProvider: defaultKubeClientProvider,
+	}
+}
+
+func defaultKubeClientProvider(context *service.ActionContext, logger *zap.SugaredLogger) (kubernetes.Client, error) {
+	kubeClient, err := kubernetes.NewKubernetesClient(context.Task.Kubeconfig, logger, &kubernetes.Config{
+		ProgressInterval: progressTrackerInterval,
+		ProgressTimeout:  progressTrackerTimeout,
+	})
+	if err != nil {
+		return nil, err
+	}
+	return kubeClient, nil
+}
+
+func (r *removeNatsOperatorStep) Execute(context *service.ActionContext, logger *zap.SugaredLogger) error {
```

Comment:
Currently, the execute logic would always render 1.X Charts/CRDs and tries to delete them, which would make sense in case of upgrading a 1.X cluster to 2.X. But if a 2.X cluster is being reconciled, the same code would try to delete 1.X Charts/CRDs again, and in my opinion that is not necessary.

Reply:
I added a TODO to the code with the reference to the feature, which makes your comment implementable, for now I think we keep it this way.